pkg/schedule/operator/operator_controller.go

```diff
@@ -512,16 +512,43 @@ func isHigherPriorityOperator(new, old *Operator) bool {
 func (oc *Controller) addOperatorInner(op *Operator) bool {
 	regionID := op.RegionID()
 
-	// If there is an old operator, replace it. The priority should be checked
-	// already.
-	if oldi, ok := oc.operators.Load(regionID); ok {
-		old := oldi.(*Operator)
-		_ = oc.removeOperatorInner(old)
-		_ = old.Replace()
-		oc.buryOperator(old)
+	old, loaded := oc.operators.LoadOrStore(regionID, op)
+	if loaded {
+		// If there is an old operator and it has lower priority, replace it
+		oldOp := old.(*Operator)
+		if !isHigherPriorityOperator(op, oldOp) {
+			log.Debug("operator already exists with higher or equal priority",
+				zap.Uint64("region-id", regionID),
+				zap.Reflect("old", oldOp),
+				zap.Reflect("new", op))
+			operatorCounter.WithLabelValues(op.Desc(), "redundant").Inc()
+			return false
+		}
+		// replace old operator
+		if !oc.operators.CompareAndSwap(regionID, oldOp, op) {
+			log.Debug("operator changed during replace, skip this add",
+				zap.Uint64("region-id", regionID),
+				zap.Reflect("old", oldOp),
+				zap.Reflect("new", op))
+			return false
+		}
+		oc.counts.dec(oldOp.SchedulerKind())
+		oc.counts.inc(op.SchedulerKind())
+		oc.ack(oldOp)
+		if oldOp.Kind()&OpMerge != 0 {
+			oc.removeRelatedMergeOperator(oldOp)
+		}
+		_ = oldOp.Replace()
+		oc.buryOperator(oldOp)
```

Comment:
Do we need to call op.Start() for new op here?

Reply:
nice catch